### ISSUE TITLE
Improve Node Utilization by Avoiding "safe-to-evict" Annotation for Druid-Managed Pods

### ISSUE THREAD

**Issue by vlerenc** [gardener/etcd-druid#766]
# How to categorize this issue?
/kind enhancement

# Stories
- As operator, I want to run my ETCDs efficiently (using the bin-packing kube-scheduler and the cluster-autoscaler down-scaling), but etcd-main(-compactor) pods disallow cluster-autoscaler down-scaling

## What would you like to be added
- [x] Do not add the `cluster-autoscaler.kubernetes.io/safe-to-evict: "false"` annotation to the etcd-main-compactor jobs (just because it's in the etcd-main spec)
- [x] Do not add the `cluster-autoscaler.kubernetes.io/safe-to-evict: "false"` annotation to the etcd-main pods for the clustered etcd-main statefulset
- [x] ~~Either switch all clusters to clustered etcd-main (except for evaluation clusters) and remove the annotation in general~~ (cost-prohibitive)
- [ ] ...or consider a more elaborate approach how to deal with the singleton etcd-main statefulset:
  - If inside the maintenance time window (with a jitter, so that we do not terminate pods all at once at the full hour, which may lead to prolonged detach/attach times)
  - If the cluster does not have the purpose "production"
  - If the requests utilization (CPU or memory) for the node hosting the pod is below 70% (atm.; ideally configurable)

  Implementation details:
  - Instead of the druid reading the purpose, gardenlet could do that and set a new field at the etcd resource that permits or forbids voluntary evictions.
  - A new controller, let's call it terminator (:-)), could do the above as part of druid.

# Motivation (Why is this needed?)
We have adopters seeing 80-90 % requests utilization on the regular seed cluster nodes, but only 40-50% for their dedicated ETCD seed cluster nodes.

**Comment by ishan16696** [comment]
> Do not add the `cluster-autoscaler.kubernetes.io/safe-to-evict: "false"` annotation to the etcd-main pods for the clustered etcd-main statefulset

When evicting a member pod from a clustered etcd, there are a few points to keep in mind:
1. All members of the etcd cluster should be functioning correctly. If one etcd member is already down in a 3 member etcd cluster, evicting another member could lead to a quorum loss. While Pod Disruption Budgets (PDBs) will prevent etcd from losing quorum, I still believe it would be better to have a precondition that check all members of the etcd cluster are healthy.
2. When evicting a member from the etcd cluster, it's preferable to choose a follower rather than the leader. This helps to avoid unnecessary leader elections.

 > Terminate the singleton etcd-main statefulset pod

To avoid a worst case scenario i.e **long restoration**, I would suggest let the backup-restore takes the schedule full snapshot which is also performed during the maintenance window before evicting a single node etcd-main pod.

**Comment by vlerenc** [comment]
It will probably not amount to much/anything, but (1) could be done right away (compactor jobs), right?

@ishan16696 How would you suggest to do that? This is about cost savings, cluster-autoscaler, evictions, PDBs. You do not have a real hook into these things, so in order for the above to be implemented, you would need to take control of everything (put the PDB to 3 minAvailable / 0 maxUnavailable and let the druid do the job if it sees the "signs"), right?

But you already run ETCD with a PDB of 2 minAvailable / 1 maxUnavailable (for clustered ETCD), so you as a team already decided to go with a flexible PDB, have you not?

**Comment by shreyas-s-rao** [comment]
> But you already run ETCD with a PDB of 2 minAvailable / 1 maxUnavailable (for clustered ETCD), so you as a team already decided to go with a flexible PDB, have you not?

@vlerenc Yes, but what does flexible mean here? The PDB allows for atmost 1 pod to be down/evicted. So the PDB should most certainly block a pod eviction if another pod is already healthy, shouldn't it?

> I still believe it would be better to have a precondition that check all members of the etcd cluster are healthy.

@ishan16696 but where would this be precondition be? Druid doesn't have any say in eviction of an etcd pod or draining of a node. Those events can occur at any time, And druid really can't block it. Well, maybe it's possible, if it puts a finalizer on the pods themselves, and wouldn't remove the finalizer unless it allows the pod to be evicted. This can possibly be another part of the scope for druid-controlled pod-updates maybe? @unmarshall 

> When evicting a member from the etcd cluster, it's preferable to choose a follower rather than the leader. This helps to avoid unnecessary leader elections.

@ishan16696 if the node being scaled down is the one hosting the leader, then there would definitely be a leader election, no doubt. But the optimisation here would be for druid to manually move leadership from this pod to another, and then unblock the pod for deletion.

**Comment by vlerenc** [comment]
> Yes, but what does flexible mean here? The PDB allows for atmost 1 pod to be down/evicted. So the PDB should most certainly block a pod eviction if another pod is already healthy, shouldn't it?

@shreyas-s-rao What I mean is that you already allow for 1 voluntary eviction (for clustered ETCDs). So, you made this decision already and are OK with it/that it may happen (in general). All I suggested, because I saw the/that PDB, was to drop the CA annotation that only influences CA. You already OKed a voluntary evictions in general in the PDB.

Or in other words: If there are general concerns with voluntary evictions (what PDBs do - which would be respected during scale-in of a node up until the timeout), why did we define a PDB allowing a voluntary eviction of 1?

**Comment by shreyas-s-rao** [comment]
Yes, I would guess so. Without this flexibility of allowing 1 replica to be evicted, it would be a nightmare to scale down any nodes at all, irrespective of the CA annotation.

**Comment by unmarshall** [comment]
>Do not add the cluster-autoscaler.kubernetes.io/safe-to-evict: "false" annotation to the etcd-main-compactor jobs (just because it's in the etcd-main spec)
> Do not add the cluster-autoscaler.kubernetes.io/safe-to-evict: "false" annotation to the etcd-main pods for the clustered etcd-main statefulset

For a multi-node etcd cluster there is no need for this annotation to be present since we already have a PDB defined which allows 1 eviction at the most. Today for a single node etcd cluster we set `minAvailable` to 0 so the annotation today prevents a single pod from being evicted and causing a downtime so here we need to be careful.

>But the optimisation here would be for druid to manually move leadership from this pod to another, and then unblock the pod for deletion.

Do we really need this? Leader election can happen at any time during a lifecycle of an etcd cluster. So how is leader election caused due to CA scaling down a node any different?

**Comment by shreyas-s-rao** [comment]
In an out-of-band discussion between myself, @unmarshall , @ishan16696 , @seshachalam-yv and @renormalize , we concluded the following: 
1. It makes sense to remove annotation `cluster-autoscaler.kubernetes.io/safe-to-evict: false` for multi-node etcds. This change needs to be done on gardenlet side when it creates the `Etcd` resource, since druid simply sees the annotation and sets it on the pod.
2. Druid should not bother about terminating pods, since it has no knowledge about shoot purpose or shoot maintenance windows. Instead, gardenlet should be the one to take care of this, since it is the one with full knowledge about this. It is not semantically correct for druid to voluntarily delete perfectly well-running pods.
  i. Gardenlet can do this by assessing nodes that are stuck in drain due to a blocking pod that has the annotation `cluster-autoscaler.kubernetes.io/safe-to-evict: false` set on it, and remove such a pod. Or better yet, it should remove the annotation during the maintenance window and add it back at the end of the maintenance window, depending on shoot purpose. Druid's job is to simply reconcile the etcd and add/update/delete annotations based on what is set in the `Etcd` resource.
3. Druid could still provide a trigger point for gardenlet to trigger deletion of a pod, like a PodDeletionTask, but this might be an overkill for a simple task of just deleting a pod.

@vlerenc PTAL.

**Comment by vlerenc** [comment]
> Gardenlet can do this by assessing nodes that are stuck in drain due to a blocking pod that has the annotation cluster-autoscaler.kubernetes.io/safe-to-evict: false set on it, and remove such a pod.

Why would that happen? The annotation only influences the CA, nothing else, no?

**Comment by voelzmo** [comment]
Just a side note on this topic:
>> I still believe it would be better to have a precondition that check all members of the etcd cluster are healthy.

>So i checked the way disruption controller computes available pods. It checks the ready condition of the Pod to determine. 

This is true, until you set [`unhealthyPodEvictionPolicy: AlwaysAllow`](https://kubernetes.io/blog/2023/01/06/unhealthy-pod-eviction-policy-for-pdbs/) in your PDB. This was specifically introduced to help get out of situations where you _do want_ eviction to happen, _although_ Pods are currently unhealthy, for example when an OOM occurs.

**Comment by unmarshall** [comment]
> This is true, until you set [unhealthyPodEvictionPolicy: AlwaysAllow](https://kubernetes.io/blog/2023/01/06/unhealthy-pod-eviction-policy-for-pdbs/) in your PDB.

Thanks for sharing this as i was unaware of this feature. Unhealthy translates to ready condition being false. For an etcd cluster if there is a loss of quorum then the ready condition for all pods will change to false. This then allows even the healthier pod (1 out of 3) to also be evicted. So in case of a permanent quorum loss this would not have much side effect. But i wondering how will this impact transient quorum loss situations.

**Comment by shreyas-s-rao** [comment]
PDB behavior and CA behavior are meant to tackle two different problems entirely. One aims to maintain minimum desired replicas for multi-node etcds, in order to keep them quorate, irrespective of what state the underlying nodes are in. Hence PDBs are necessary for multi-node etcds, and CA does not adversely affect the quorum of a multi-node etcd. Now, setting the field `unhealthyPodEvictionPolicy` for the PDB here can easily lead to downtimes for the etcd, especially during node rollout, because now we allow the PDB to go ahead and allow eviction of more than `minAvailable` number of pods. This seems counter-intuitive.

On the other hand, CA aims to control node lifecycle, including draining of nodes that run single-node etcds on them, and expects that an etcd pod should not block the draining of a node indefinitely. PDBs play no part for single-node etcds, since a single-node etcd pod must be rolled sometime or another, leading to a temporary unavailability of the etcd. Here, setting the `unhealthyPodEvictionPolicy` makes little difference, because at the end of the day, the main issue CA faces is that a healthy single-node etcd pod running on a node cannot be evicted because of the annotation `cluster-autoscaler.kubernetes.io/safe-to-evict` set on it, so PDB has no part to play here. Instead, it should be the responsibility of the one who blocked the pod eviction (gardenlet) to unblock pod eviction during node drain, and then re-block it after the pod has been relocated to a new, already-rolled node.

**Comment by shreyas-s-rao** [comment]
Meeting minutes from an out-of-band discussion between myself, @vlerenc, @unmarshall , @ishan16696 , @renormalize , @seshachalam-yv, @ashwani2k  :
- Agreed to remove `cluster-autoscaler.kubernetes.io/safe-to-evict: "false"` for compaction job pods, as suggested by @vlerenc 
- Agreed to remove `cluster-autoscaler.kubernetes.io/safe-to-evict: "false"` for multi-node etcd pods, as suggested by @vlerenc 
- For single-node etcd pods, we discussed the following:
  - CA does not select a node for drain if the node runs at least one pod with the annotation `cluster-autoscaler.kubernetes.io/safe-to-evict: "false"` set on it. So there is no deterministic way for an agent to figure out if a node is "to be" drained in order to decide to kick the etcd pods off of it during their respective maintenance windows. The hope is that kicking the etcd pods will allow it to be scheduled to a node with larger utilisation, thanks to bin-packing.
  - Another option is to set a PDB with `minAvailable: 1` for single-node etcds and change this to `minAvailable: 0` during maintenance window for each etcd. But CA considers PDB disruptions for the pods running on the node while selecting the node to drain. So even with this approach, we face the same issue of non-overlapping maintenance windows, and having to select the node to drain and then kick the etcd pods off one by one during their respective maintenance windows. Additionally, the PDB would block MCM from draining nodes during node rollouts, effectively blocking them or forcing MCM to forcefully shut down machines.
  - Selecting a node for drain should be done in such a way that it doesn't interfere with CA's criteria for selecting a node for drain, based on node usage. For instance, if CA is configured to drain a node which is only 20% utilised, then our custom agent needs to select the node when it is at 40% usage, taints the node, and then starts kicking etcd pods off voluntarily during their maintenance windows.
  - @unmarshall suggested co-locating etcd pods with same/similar maintenance windows onto the same nodes, so that the maintenance windows for the etcds on a node to be drained would more or less fall into the same time period, allowing for an agent to remove the safe-to-evict annotation and kick all the etcd pods off at the same time.
  - Another suggestion is to entirely move to multi-node etcd for `etcd-main`, even for non-HA control planes, and entirely get rid of `cluster-autoscaler.kubernetes.io/safe-to-evict: "false"`, possibly based on shoot purpose (testing, production, etc). This way, we allow CA to drain nodes from etcd worker pools, without any single-node etcd pods to block them.

Action items:
- @vlerenc to check stats for switching from single-node to multi-node etcds for non-HA control planes, and check if cost savings with nodes would justify the extra costs for running 2 extra replicas of etcd-main (memory, CPU, disk). The main cost differential seems to be with etcd disks, especially for AWS:
  - New AWS etcds pods will soon be created with `25Gi` volumes instead of the previous `80Gi` value , via https://github.com/gardener/gardener-extension-provider-aws/pull/933
  - Reducing volume size for old etcd pods will require https://github.com/gardener/etcd-druid/issues/481 to be implemented, allowing druid to roll volumes
